compiler/rustc_mir/src/transform/pass.rs

```diff
@@ -0,0 +1,136 @@
+use std::borrow::Cow;
+
+use rustc_middle::mir::{Body, MirPhase};
+use rustc_middle::ty::TyCtxt;
+
+use super::{dump_mir, validate};
+
+/// A streamlined trait that you can implement to create a pass; the
+/// pass will be named after the type, and it will consist of a main
+/// loop that goes over each available MIR and applies `run_pass`.
+pub trait MirPass<'tcx> {
+    const LEVEL: OptLevel;
+
+    fn name(&self) -> Cow<'_, str> {
+        let name = std::any::type_name::<Self>();
+        name.rsplit(':').next().unwrap().into()
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>);
+}
+
+const UNSOUND_MIR_OPT_LEVEL: u8 = 2;
+
+pub enum OptLevel {
+    /// Passes that will run at `-Zmir-opt-level=N` or higher.
+    N(u8),
+
+    /// Passes that are known or suspected to cause miscompilations.
+    ///
+    /// These passes only run if `-Zunsound-mir-opts` is enabled and we are at `mir-opt-level=2` or
+    /// above.
+    Unsound,
```

Comment:
I think the original idea was that `unsound-mir-opts` would be completely orthogonal to the optimization level. I could imagine us having an optimization that's buggy (thus unsound) but improves compilation performance quite a bit so we'd eventually want it to run at the default opt level.

I'm not sure this matters in practice though and it would probably complicate things a bit more to make the code support that when we don't currently need it.

Reply:
For the record, one unsound MIR opt runs at opt-level 2 or above, the other runs at opt-level 1 or above. I also don't think this matters in practice, although I've switched to running unsound opts at opt-level 1 or above since it made blessing tests slightly easier.